python/python/lance/fragment.py

```diff
@@ -37,53 +36,161 @@
+    id : int
+        The ID of the fragment.
+    files : List[DataFile]
+        The data files of the fragment. Each data file must have the same number
+        of rows. Each file stores a different subset of the columns.
+    physical_rows : int
+        The number of rows originally in this fragment. This is the number of rows
+        in the data files before deletions.
+    deletion_file : Optional[DeletionFile]
+        The deletion file, if any.
+    row_id_meta : Optional[RowIdMeta]
+        The row id metadata, if any.
+    """
 
-    def __repr__(self):
-        return self._metadata.__repr__()
+    id: int
+    files: List[DataFile]
+    physical_rows: int
+    deletion_file: Optional[DeletionFile] = None
+    row_id_meta: Optional[RowIdMeta] = None
 
-    def __reduce__(self):
-        return (FragmentMetadata, (self._metadata.json(),))
+    @property
+    def num_deletions(self) -> int:
+        """The number of rows that have been deleted from this fragment."""
+        if self.deletion_file is None:
+            return 0
+        else:
+            return self.deletion_file.num_deleted_rows
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, FragmentMetadata):
-            return False
-        return self._metadata.__eq__(other._metadata)
+    @property
+    def num_rows(self) -> int:
+        """The number of rows in this fragment after deletions."""
+        return self.physical_rows - self.num_deletions
 
-    def to_json(self) -> str:
-        """Serialize :class:`FragmentMetadata` to a JSON blob"""
-        return json.loads(self._metadata.json())
+    def data_files(self) -> List[DataFile]:
+        warnings.warn(
+            "FragmentMetadata.data_files is deprecated. Use .files instead.",
+            DeprecationWarning,
+        )
+        return self.files
+
+    def to_json(self) -> dict:
+        """Get this as a simple JSON-serializable dictionary."""
+        files = [asdict(f) for f in self.files]
+        for f in files:
+            f["path"] = f.pop("_path")
+        return dict(
+            id=self.id,
+            files=files,
+            physical_rows=self.physical_rows,
+            deletion_file=(
+                self.deletion_file.asdict() if self.deletion_file is not None else None
+            ),
+            row_id_meta=(
+                self.row_id_meta.asdict() if self.row_id_meta is not None else None
+            ),
+        )
 
     @staticmethod
     def from_json(json_data: str) -> FragmentMetadata:
-        """Reconstruct :class:`FragmentMetadata` from a JSON blob"""
-        return FragmentMetadata(json_data)
+        json_data = json.loads(json_data)
+
+        deletion_file = json_data.get("deletion_file")
+        if deletion_file is not None:
+            deletion_file = DeletionFile(**deletion_file)
+
+        row_id_meta = json_data.get("row_id_meta")
+        if row_id_meta is not None:
+            row_id_meta = RowIdMeta(**row_id_meta)
```

Comment:
I think I've fallen behind the times again.  What is `row_id_meta`?  Is this where we put a row id mapping file for stable row id stuff?

Reply:
Yeah, each fragment has it's own row id index stored here. Either an inline index (if it's small), or a reference to a separate file.